airbyte-cdk/python/airbyte_cdk/sources/file_based/file_types/unstructured_parser.py

```diff
@@ -0,0 +1,123 @@
+#
+# Copyright (c) 2023 Airbyte, Inc., all rights reserved.
+#
+import logging
+from io import BytesIO, IOBase
+from typing import Any, Dict, Iterable, List, Mapping, Optional
+
+from airbyte_cdk.sources.file_based.config.file_based_stream_config import FileBasedStreamConfig
+from airbyte_cdk.sources.file_based.exceptions import FileBasedSourceError, RecordParseError
+from airbyte_cdk.sources.file_based.file_based_stream_reader import AbstractFileBasedStreamReader, FileReadMode
+from airbyte_cdk.sources.file_based.file_types.file_type_parser import FileTypeParser
+from airbyte_cdk.sources.file_based.remote_file import RemoteFile
+from airbyte_cdk.sources.file_based.schema_helpers import SchemaType
+
+
+class UnstructuredParser(FileTypeParser):
+    @property
+    def parser_max_n_files_for_schema_inference(self) -> Optional[int]:
+        """
+        Just check one file as the schema is static
+        """
+        return 1
+
+    @property
+    def parser_max_n_files_for_parsability(self) -> Optional[int]:
+        """
+        Do not check any files for parsability because it might be an expensive operation and doesn't give much confidence whether the sync will succeed.
+        """
+        return 0
+
+    async def infer_schema(
+        self,
+        config: FileBasedStreamConfig,
+        file: RemoteFile,
+        stream_reader: AbstractFileBasedStreamReader,
+        logger: logging.Logger,
+    ) -> SchemaType:
+        with stream_reader.open_file(file, self.file_read_mode, None, logger) as file_handle:
+            filetype = self._get_filetype(file_handle, file.uri)
+
+            if filetype not in self._supported_file_types():
+                raise RecordParseError(FileBasedSourceError.ERROR_PARSING_RECORD, filename=file.uri)
+
+            return {
+                "content": {"type": "string"},
+                "document_key": {"type": "string"},
+            }
+
+    def parse_records(
+        self,
+        config: FileBasedStreamConfig,
+        file: RemoteFile,
+        stream_reader: AbstractFileBasedStreamReader,
+        logger: logging.Logger,
+        discovered_schema: Optional[Mapping[str, SchemaType]],
+    ) -> Iterable[Dict[str, Any]]:
+        with stream_reader.open_file(file, self.file_read_mode, None, logger) as file_handle:
+            markdown = self._read_file(file_handle, file.uri)
+            yield {
+                "content": markdown,
+                "document_key": file.uri,
+            }
+
+    def _read_file(self, file_handle: IOBase, file_name: str) -> str:
+        from unstructured.file_utils.filetype import FileType
+        from unstructured.partition.auto import partition
+        from unstructured.partition.md import optional_decode
```

Comment:
Regarding these inline imports, are these not on the top of the file because they are extras in the CDK?

I think the general best practice for these scenarios would be something like this at the top of the file...

```
try:
    from unstructured.file_utils.filetype import FileType
    from unstructured.partition.auto import partition
    from unstructured.partition.md import optional_decode
except ImportError as ex:
    logger.warning(f"Unstructured libraries could not be imported. {ex}")
    FileType = None
    partition = None
    optional_decode = None
```

Reply:
They are inline because they are quite slow - if they are on the top of the file it would slow down the connector even if the parser is not used. I can add a comment to explain this (or maybe there is a more elegant way to handle this I’m not aware of)

Reply:
In that case, what do you think about putting into a helper function called by the class constructor?

Pseudocode, may need adjustment:

```py
...
unstructured_lib = None

def _import_unstructured():
    """Dynamically imported as needed, due to slow import speed."""
    global unstructured_lib

    import unstructured
    unstructured_lib = unstructured

class UnstructuredParser(FileTypeParser):
    def __init__():
        _import_unstructured()
```